ENH add `only_non_negative` parameter to `_check_sample_weight`

simonandras opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
Fixes #20036. 
Partially addressed https://github.com/scikit-learn/scikit-learn/issues/15531

#### What does this implement/fix? Explain your changes.
The LinearRegression() objects .fit() method gives unclear error message if the sample_weight parameter of the fit() method has negative value. The exception occured only at the call of linalg.lstsq(X, y). I think its better if the negative value check is done at the beginning of the .fit() method. The new message is: "The weights of the samples cannot be negative".

#### Any other comments?
I think the check for negative values should not be done inside the _check_sample_weight() function but inside the given .fit() method. Maybe there are other cases too where this exception message and the negative weight value check can be useful.

This is my first PR to the open source community and i wellcome any comment or advice related to this PR!


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

simonandras committed:
Adding informative error message for the negative weight case (linear regression)

glemaitre commented:
@simonandras Could you check the message that I let on the original issue: https://github.com/scikit-learn/scikit-learn/issues/20036#issuecomment-910271359

I think that we should change `_check_sample_weight` instead.

simonandras committed:
The negativity check is now in _check_sample_weight which can be turned on with the only_positive bool parameter (beckward compatible). For the check the alredy written check_non_negative is used.

simonandras committed:
Adding parameter documentation to the changed _check_sample_weight and rename the only_positive parameter to only_non_negative, because zeros are accepted.

simonandras commented:
Thank you for the reply!
I changed the _check_sample_weight function as you suggested. Now i used the alredy writen check_non_negative function for it. This function has an error message writer part, only the "whom" parameter had to be added. Also added the only_non_negative parameter to  _check_sample_weight (insted of only_positive, because zeros are allowed).

simonandras committed:
Making some lines shorter than 88 char to pass tests

simonandras committed:
Make one line documentation in two lines to pass test (it was too long)

simonandras committed:
Delete whitespace in blank line

simonandras committed:
Add a dot to end of sentence (in modified function documentation)

simonandras committed:
formatting with black .

simonandras commented:
I wasnt able to make all checks succesfull. I think the problem is with the test_adaboost_negative_weight_error function and the new error message but im not sure what the problem actually is and what to do with it. Any help would be appriciated!

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -1495,6 +1497,9 @@ def _check_sample_weight(sample_weight, X, dtype=None, copy=False):
     X : {ndarray, list, sparse matrix}
         Input data.
 
+    only_non_negative : if True then a non negativity check for the sample_weight
+        will be done.
```
```suggestion
    only_non_negative : bool, default=False,
        Whether or not the weights are expected to be non-negative.
```

glemaitre commented on sklearn/linear_model/_base.py:
```diff
@@ -662,8 +662,10 @@ def fit(self, X, y, sample_weight=None):
 
-        if sample_weight is not None:
-            sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
+        if sample_weight is not None:  # the weights should be non-negative
```
There is no need for this comment since passing the keyword argument is self-explanatory.

```suggestion
        if sample_weight is not None:
```

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -1540,6 +1545,9 @@ def _check_sample_weight(sample_weight, X, dtype=None, copy=False):
             )
 
+    if only_non_negative:
+        check_non_negative(sample_weight, "sample weight")
```
```suggestion
        check_non_negative(sample_weight, "`sample_weight`")
```

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -1479,7 +1479,9 @@ def _check_psd_eigenvalues(lambdas, enable_warnings=False):
 
-def _check_sample_weight(sample_weight, X, dtype=None, copy=False):
+def _check_sample_weight(
+    sample_weight, X, only_non_negative=False, dtype=None, copy=False
```
Here there is a silent bug. You need to put `only_non_negative` as the last argument because some call of `_check_sample_weight` are using positional argument. Thus, passing previously the dtype as a positional argument become the value of `only_non_negative`.

glemaitre commented:
Here I give you a patch that show as well how to modify the current code such that the test failures should be solved:

```diff
diff --git a/sklearn/ensemble/_weight_boosting.py b/sklearn/ensemble/_weight_boosting.py
index 77ef449ba1..ed3b0ad853 100644
--- a/sklearn/ensemble/_weight_boosting.py
+++ b/sklearn/ensemble/_weight_boosting.py
@@ -123,10 +123,10 @@ class BaseWeightBoosting(BaseEnsemble, metaclass=ABCMeta):
             y_numeric=is_regressor(self),
         )
 
-        sample_weight = _check_sample_weight(sample_weight, X, np.float64, copy=True)
+        sample_weight = _check_sample_weight(
+            sample_weight, X, np.float64, copy=True, non_negative=True
+        )
         sample_weight /= sample_weight.sum()
-        if np.any(sample_weight < 0):
-            raise ValueError("sample_weight cannot contain negative weights")
 
         # Check parameters
         self._validate_estimator()
@@ -136,7 +136,7 @@ class BaseWeightBoosting(BaseEnsemble, metaclass=ABCMeta):
         self.estimator_weights_ = np.zeros(self.n_estimators, dtype=np.float64)
         self.estimator_errors_ = np.ones(self.n_estimators, dtype=np.float64)
 
-        # Initializion of the random number instance that will be used to
+        # Initialization of the random number instance that will be used to
         # generate a seed at each iteration
         random_state = check_random_state(self.random_state)
 
diff --git a/sklearn/ensemble/tests/test_weight_boosting.py b/sklearn/ensemble/tests/test_weight_boosting.py
index 6927d47c11..159f83abf2 100755
--- a/sklearn/ensemble/tests/test_weight_boosting.py
+++ b/sklearn/ensemble/tests/test_weight_boosting.py
@@ -576,6 +576,6 @@ def test_adaboost_negative_weight_error(model, X, y):
     sample_weight = np.ones_like(y)
     sample_weight[-1] = -10
 
-    err_msg = "sample_weight cannot contain negative weight"
+    err_msg = "Negative values in data passed to `sample_weight`"
     with pytest.raises(ValueError, match=err_msg):
         model.fit(X, y, sample_weight=sample_weight)
diff --git a/sklearn/linear_model/_base.py b/sklearn/linear_model/_base.py
index 6e1d6acb0c..514525a14a 100644
--- a/sklearn/linear_model/_base.py
+++ b/sklearn/linear_model/_base.py
@@ -662,7 +662,7 @@ class LinearRegression(MultiOutputMixin, RegressorMixin, LinearModel):
             X, y, accept_sparse=accept_sparse, y_numeric=True, multi_output=True
         )
 
-        if sample_weight is not None:  # the weights should be non-negative
+        if sample_weight is not None:
             sample_weight = _check_sample_weight(
                 sample_weight, X, only_non_negative=True, dtype=X.dtype
             )
diff --git a/sklearn/neighbors/_kde.py b/sklearn/neighbors/_kde.py
index 328a13371b..dcfc7b51f4 100644
--- a/sklearn/neighbors/_kde.py
+++ b/sklearn/neighbors/_kde.py
@@ -191,9 +191,7 @@ class KernelDensity(BaseEstimator):
         X = self._validate_data(X, order="C", dtype=DTYPE)
 
         if sample_weight is not None:
-            sample_weight = _check_sample_weight(sample_weight, X, DTYPE)
-            if sample_weight.min() <= 0:
-                raise ValueError("sample_weight must have positive values")
+            sample_weight = _check_sample_weight(sample_weight, X, DTYPE, only_non_negative=True)
 
         kwargs = self.metric_params
         if kwargs is None:
diff --git a/sklearn/neighbors/tests/test_kde.py b/sklearn/neighbors/tests/test_kde.py
index 84f7623c8d..d4fb775c44 100644
--- a/sklearn/neighbors/tests/test_kde.py
+++ b/sklearn/neighbors/tests/test_kde.py
@@ -209,7 +209,7 @@ def test_sample_weight_invalid():
     data = np.reshape([1.0, 2.0, 3.0], (-1, 1))
 
     sample_weight = [0.1, -0.2, 0.3]
-    expected_err = "sample_weight must have positive values"
+    expected_err = "Negative values in data passed to `sample_weight`"
     with pytest.raises(ValueError, match=expected_err):
         kde.fit(data, sample_weight=sample_weight)
 
diff --git a/sklearn/utils/validation.py b/sklearn/utils/validation.py
index 554f821539..eb015d287d 100644
--- a/sklearn/utils/validation.py
+++ b/sklearn/utils/validation.py
@@ -1480,7 +1480,11 @@ def _check_psd_eigenvalues(lambdas, enable_warnings=False):
 
 
 def _check_sample_weight(
-    sample_weight, X, only_non_negative=False, dtype=None, copy=False
+    sample_weight,
+    X,
+    dtype=None,
+    copy=False,
+    only_non_negative=False,
 ):
     """Validate sample weights.
 
@@ -1497,9 +1501,6 @@ def _check_sample_weight(
     X : {ndarray, list, sparse matrix}
         Input data.
 
-    only_non_negative : if True then a non negativity check for the sample_weight
-        will be done.
-
     dtype : dtype, default=None
        dtype of the validated `sample_weight`.
        If None, and the input `sample_weight` is an array, the dtype of the
@@ -1510,6 +1511,9 @@ def _check_sample_weight(
     copy : bool, default=False
         If True, a copy of sample_weight will be created.
 
+    only_non_negative : bool, default=False,
+        Whether or not the weights are expected to be non-negative.
+
     Returns
     -------
     sample_weight : ndarray of shape (n_samples,)
@@ -1546,7 +1550,7 @@ def _check_sample_weight(
             )
 
     if only_non_negative:
-        check_non_negative(sample_weight, "sample weight")
+        check_non_negative(sample_weight, "`sample_weight`")
 
     return sample_weight
```

glemaitre commented:
Indeed, you need to:

- move `only_non_negative` as the last parameter of the `_check_sample_weight` signature.
- modify AdaBoost and KDE such that we don't check anymore by hand for negative weights.
- change the error message in the associated test files.

Then everythin should be fine :)

glemaitre commented:
In addition, we will want a test in `test_validation.py` to check that we raise the expected error when using `_check_sample_weigth`.
We don't need an entry in the changelog since the tool is part of the private API (probably dev API thouht).

simonandras committed:
Update sklearn/utils/validation.py

update documentation

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

simonandras committed:
Update sklearn/linear_model/_base.py

delete unnecessary comment

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

simonandras committed:
Update sklearn/utils/validation.py

Add `...` to function name in the error message

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

simonandras committed:
Merge branch 'scikit-learn:main' into my-new-branch

simonandras committed:
Merge branch 'main' of https://github.com/simonandras/scikit-learn into my-new-branch

simonandras committed:
Merge branch 'my-new-branch' of https://github.com/simonandras/scikit-learn into my-new-branch

simonandras committed:
Solve the previous problems

simonandras committed:
reformatting with black .

simonandras committed:
In the documentation of _check_sample_weight changing the 3 space indents to 4 space indents

simonandras committed:
Adding test to test_validation.py to test the negative weight error in the _check_sample_weigth function

simonandras committed:
black .

simonandras commented:
I made the changes and wrote the test inside the already exisiting test_check_sample_weight function as a new part. Every check is alright now except one, which is due to there is no new entry in the changelog. As you said we dont need an entry for this one, so this failed check can be ignored?

simonandras committed:
In LinearRegression in fit method in _check_sample_weight function call make the parameter order like in the function definition

glemaitre commented:
Please add an entry to the change log at `doc/whats_new/v1.0.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`.

simonandras committed:
Added entry to the changelog (sklearn.utils). Also corrected 2 typo in the changelog.

simonandras committed:
Merge branch 'main' into my-new-branch

glemaitre reviewed (approved):
LGTM. Probably @rth can have a look at this one.

ogrisel committed:
Merge branch 'main' into my-new-branch

ogrisel reviewed (approved):
LGTM as well!

simonandras committed:
Merge branch 'scikit-learn:main' into my-new-branch

glemaitre commented:
I will move this item for 1.1 since this is not a blocker for the release

glemaitre committed:
move whats new version

glemaitre reviewed (approved):
I moved the changelog entry. I will merge if the CIs pass
